scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryNetwork.scala

```diff
@@ -0,0 +1,398 @@
+import io.iohk.scalanet.discovery.ethereum.v4.Payload.Neighbors
+
+/** Present a stateless facade implementing the RPC methods
+  * that correspond to the discovery protocol messages on top
+  * of the peer group representing the other nodes.
+  */
+trait DiscoveryNetwork[A] extends DiscoveryRPC[DiscoveryNetwork.Peer[A]] {
+
+  /** Start handling incoming requests using the local RPC interface.
+    * The remote side is identified by its ID and address.*/
+  def startHandling(handler: DiscoveryRPC[DiscoveryNetwork.Peer[A]]): Task[CancelableF[Task]]
+}
+
+object DiscoveryNetwork {
+
+  /** The pair of node ID and the UDP socket where it can be contacted or where it contacted us from.
+    * We have to use the pair for addressing a peer as well to set an expectation of the identity we
+    * expect to talk to, i.e. who should sign the packets.
+    */
+  case class Peer[A](id: PublicKey, address: A)
+
+  // Errors that stop the processing of incoming messages on a channel.
+  class PacketException(message: String) extends Exception(message) with NoStackTrace
+
+  def apply[A](
+      peerGroup: PeerGroup[A, Packet],
+      privateKey: PrivateKey,
+      toNodeAddress: A => Node.Address,
+      config: DiscoveryConfig
+  )(implicit codec: Codec[Payload], sigalg: SigAlg, clock: Clock[Task]): Task[DiscoveryNetwork[A]] = Task {
+    new DiscoveryNetwork[A] with LazyLogging {
+
+      import DiscoveryRPC.ENRSeq
+      import Payload._
+
+      private val expirationMillis = config.messageExpiration.toMillis
+      private val maxClockDriftMillis = config.maxClockDrift.toMillis
+      private val currentTimeMillis = clock.realTime(MILLISECONDS)
+
+      private val maxNeighborsPerPacket = getMaxNeighborsPerPacket
+
+      // This is only sent in Ping packets and is basically ignored by nodes.
+      private val localNodeAddress = toNodeAddress(peerGroup.processAddress)
+
+      /** Start a fiber that accepts incoming channels and starts a dedicated fiber
+        * to handle every channel separtely, processing their messages one by one.
+        * This is fair: every remote connection can be throttled independently
+        * of each other, as well as based on operation type by the `handler` itself.
+        */
+      override def startHandling(handler: DiscoveryRPC[Peer[A]]): Task[CancelableF[Task]] =
+        for {
+          cancelToken <- Deferred[Task, Unit]
+          _ <- peerGroup
+            .nextServerEvent()
+            .withCancelToken(cancelToken)
+            .toIterant
+            .mapEval {
+              case ChannelCreated(channel, release) =>
+                handleChannel(handler, channel, cancelToken)
+                  .guarantee(release)
+                  .onErrorRecover {
+                    case ex: TimeoutException =>
+                    case NonFatal(ex) =>
+                      logger.error(s"Error handling channel from ${channel.to}: $ex")
+                  }
+                  .startAndForget
```

Comment:
As we are starting concurrent fiber for each incoming channel, shouldn't we use some semaphore to limit the concurrency ? Otherwise it is easy to use up server resources by just sending messages from different adressess.

Reply:
Great question! I wanted to look up how much resource a fiber actually takes; it looks like it's basically a `Task` scheduled to run asynchronously with an attached Promise and a cancel token. I don't know how many tasks can execute concurrently. 

But we have to do something with the incoming message here that isn't blocking the `mapEval` otherwise we'd block processing all incoming connections. I guess we could track something like the number of connections from the same IP address and immediately release if there are more than some threshold. 

The fiber itself it will be closed down after a minute of inactivity.

I thought a semaphore could be used on the higher level component to for example limit the number of certain concurrent operations, which is another kind of protection against resource exhaustion, but it's not obvious how to make it fair.

In the case of discovery, to do anything heavy like FindNode first these addresses actually have to respond to a Ping. I guess it's not a bad waste of resources to ask a node to keep pinging random ports on your machine, so if we limited the number of connections from any given IP we could at least keep that in check.

Reply:
llimiting number of concurrent incoming channels from any given ip could be probably done at peer group level, and we could try to use for it netty `AbstractRemoteAddressFilter`  (or maybe it modified version as orginal is designed for connection based transport like tcp, and not udp). 

I just wonder if we should have any limits in case of getting spammed from like 10000 or more diffent ip-s at once. Ultimatly fibers are run on underlying scheduler threads, so from user perspective this is one possibility to limit resources, but i wonder if we should't have option to limit resources usage at higher lever of abstraction like semaphore with 10k of permits (or number configurable by user) which is basically saying: that is at most concurrent connection i would like to have and block accpeting next connectionns until there will more free resources. (but again maybe it should be done at peer group level ?)

Reply:
tldr: it pretty non trivial topic lets deal with it separatly

Reply:
I agree, let's open a new ticket for it. Thank you for the pointer to `AbstractRemoteAddressFilter`, looks like a good fit indeed! 

We can have semaphores to avoid being swamped by requests from 10,000 different legit addresses, all asking us to find neighbours all the time, it's easy to say we only respond to, say, 10 FindNode requests at any time, and move on if we can't get a permit within the request timeout. We can also add rate limits per individual channel, so they can only issue so many lookups per time interval.